src/__tests__/types.spec-d.ts

```diff
@@ -0,0 +1,145 @@
+    expectTypeOf<t.UnionToIntersection<number | number>>().not.toBeAny()
+    expectTypeOf<t.UnionToIntersection<number | number>>().not.toBeUnknown()
+
+    expectTypeOf<t.UnionToIntersection<ParserNumberAndString>>().toMatchTypeOf<
+      t.FailingParser &
+        t.SucceedingParser<number> &
+        t.UnsafeParser<number> &
+        t.SucceedingParser<string> &
+        t.UnsafeParser<string>
+    >()
+  })
+
+  it('`UnionToTuplePreserving` should correctly infer right signature', () => {
+    expectTypeOf<t.UnionToTuplePreserving<number>>().toMatchTypeOf<[number]>()
+    expectTypeOf<t.UnionToTuplePreserving<NumberAndString>>().toMatchTypeOf<[number, string]>()
+  })
+
+  it('`UnwrapParserTuple` should correctly infer right signature', () => {
+    expectTypeOf<t.UnwrapParserTuple<[t.Parser<string>]>>().toMatchTypeOf<[string]>()
+
+    expectTypeOf<t.UnwrapParserTuple<ParsersTuple>>().toMatchTypeOf<[string, number]>()
+    expectTypeOf<t.UnwrapParserTuple<[t.Parser<string>, t.Parser<string>]>>().toMatchTypeOf<
+      [string, string]
+    >()
+  })
+
+  it('`TupleToUnion` should correctly infer right signature', () => {
+    expectTypeOf<t.TupleToUnion<[string, string]>>().toMatchTypeOf<string>()
+    expectTypeOf<t.TupleToUnion<[string, string]>>().not.toBeAny()
+    expectTypeOf<t.TupleToUnion<[string, string]>>().not.toBeUnknown()
+
+    expectTypeOf<t.TupleToUnion<[string, number]>>().toMatchTypeOf<string | number>()
+    expectTypeOf<t.TupleToUnion<[string, number]>>().not.toBeAny()
+    expectTypeOf<t.TupleToUnion<[string, number]>>().not.toBeUnknown()
+  })
+})
+
+describe('utilities', () => {
+  type ParsersTuple = [t.Parser<string>, t.Parser<number>, t.Parser<boolean>]
+  type NumberParsersArray = Array<t.Parser<number>>
+  type TupleResult = [string, number, boolean]
+  type ParsersUnion = t.Parser<string> | t.Parser<number> | t.Parser<boolean>
+
+  it('`ToTuple` should correctly infer right signature', () => {
+    expectTypeOf<t.ToTuple<ParsersTuple>>().toMatchTypeOf<TupleResult>()
+
+    expectTypeOf<t.ToTuple<NumberParsersArray>>().toMatchTypeOf<[]>()
+  })
+
+  it('`ToTupleOrArray` should correctly infer right signature', () => {
+    expectTypeOf<t.ToTupleOrArray<ParsersTuple>>().toMatchTypeOf<TupleResult>()
+    expectTypeOf<t.ToTupleOrArray<NumberParsersArray>>().toMatchTypeOf<number[]>()
+
+    expectTypeOf<t.ToTupleOrArray<NumberParsersArray>>().not.toMatchTypeOf<string[]>()
+  })
+
+  it('`ToUnion` should correctly infer right signature', () => {
+    expectTypeOf<t.ToUnion<ParsersTuple>>().toMatchTypeOf<string | number | boolean>()
+
+    expectTypeOf<t.ToUnion<NumberParsersArray>>().toMatchTypeOf<number>()
+    expectTypeOf<t.ToUnion<NumberParsersArray>>().not.toMatchTypeOf<string>()
+  })
+
+  it('`UnwrapUnion` should correctly infer right signature', () => {
+    expectTypeOf<t.UnwrapUnion<ParsersUnion>>().toMatchTypeOf<
+      [number, string, boolean] | [string, number, boolean]
+    >()
```

Comment:
This (and pretty much any place where we have tuples AFAIK) was flaky for me locally. Usually it should be just `[string, number, boolean]` but sometimes it shifts to `[number, string, boolean]`. Idk where it's happening and why, so please be careful and if it's not reproducible by deleting the first part (`[number, string, boolean]`) we can probably ignore it.

Reply:
TS for some reason re-orders union members, so, yeah, we may have different results sometimes, but since this type helper (`UnwrapUnion`) is used only for turning this:

```ts
Parser<string> | Parser<number> | ...
```

Into that:

```ts
Parser<string | number | ...>
```

And essentially we don't care about the order of union members, we are good.